### ISSUE TITLE
Shade ANTLR to avoid version conflicts with other packages

### ISSUE THREAD

**Issue by bestbeforetoday** [substrait-io/substrait-java#259]
ANTLR appears to be particularly prone to breakages due to version differences. When substrait-core is used with other packages that also have dependencies on ANTLR (such as Spark), differences in ANTLR dependency versions frequently cause breakages.

For example, depending on which dependency version is picked up first:
```
ANTLR Tool version 4.9.3 used for code generation does not match the current runtime version 4.13.1
ANTLR Runtime version 4.9.3 used for parser compilation does not match the current runtime version 4.13.1
```

or:
```
ANTLR Tool version 4.13.1 used for code generation does not match the current runtime version 4.9.3
ANTLR Runtime version 4.13.1 used for parser compilation does not match the current runtime version 4.9.3
```

I propose solving this issue by shading the ANTLR dependency into substrait-core using the [Gradle shadow plugin](https://imperceptiblethoughts.com/shadow/configuration/relocation/). An example of prior art for this approach can be found in [this issue](https://issuetracker.google.com/issues/150106190) and its corresponding [code change](https://android-review.googlesource.com/1259721).

**Comment by bestbeforetoday** [comment]
If you agree with the approach, I would be happy to try to deliver a fix.

**Comment by Blizzara** [comment]
I at least totally agree with the issue - and already made a PR for the fix: https://github.com/substrait-io/substrait-java/pull/258 ;) Would just need to get it reviewed

**Comment by bestbeforetoday** [comment]
@Blizzara I gave your change a quick try locally but unfortunately it didn't build for me. First a formatting violation and then — when I fixed the formatting — 463 test failures due to NoClassDefFoundError.

Since I made a similar change yesterday, I've raised that in PR #260. Any additional improvements are very welcome!

**Comment by Blizzara** [comment]
Yeah I see, I fixed the formatting but Isthmus tests fail due to NoClassDefFoundError. Adding `jar { finalizedBy(shadowJar) }` would probably help but then gradle complains because the dependencies are not handled correctly...

**Comment by Blizzara** [comment]
I pushed a version of my PR which should now work, though I find the way I got it to run a bit questionable. Happy to defer to yours if you find something better!

**Comment by bestbeforetoday** [comment]
Released in [v0.32.0](https://github.com/substrait-io/substrait-java/releases/tag/v0.32.0).
